Firmware/SwitchPro/Joystick.c

```diff
@@ -0,0 +1,420 @@
+            switches[i].state = 0;
+            if(switches[i].state != switches[i].lastReport) {
+                switchesChanged = 1;
+            }
+        }
+    }
+}
+
+void nunchuck_init(void) {
+    // try to say hello
+    if(!i2c_start(NUNCHUCK_ADDR | I2C_WRITE)) {
+        i2c_write(0xF0);
+        i2c_write(0x55);
+        i2c_stop();
+        _delay_ms(25);
+        
+        i2c_start(NUNCHUCK_ADDR | I2C_WRITE);
+        i2c_write(0xFB);
+        i2c_write(0x00);
+        i2c_stop();
+        _delay_ms(25);
+        nunchuck_online();
+    } else {
+        nunchuck_offline();
+    }
+}
+
+uint8_t nunchuck_readByte(uint8_t address) {
+    uint8_t data = 0xFF;
+    
+    if(!nunchuckReady) {
+        nunchuck_init();
+    }
+
+    if(!i2c_start(NUNCHUCK_ADDR | I2C_WRITE)) {
+        i2c_write(address);
+        i2c_stop();
+
+        i2c_start(NUNCHUCK_ADDR | I2C_READ);
+        data = i2c_readNak();
+        i2c_stop();
+        nunchuck_online();
+    } else {
+        nunchuck_offline();
+    }
+    return data;
+}
+
+// Starting at address, read n bytes and return the last
+void nunchuck_readMany(uint8_t address, uint8_t *data, uint8_t count) {
+    
+    if(!nunchuckReady) {
+        nunchuck_init();
+    }
+
+    if(!i2c_start(NUNCHUCK_ADDR | I2C_WRITE)) {
+        i2c_write(address);
+        i2c_stop();
+
+        i2c_start(NUNCHUCK_ADDR | I2C_READ);
+        for(uint8_t i = 0; i < count-1; i++) {
+            data[i] = i2c_readAck();
+        }
+        data[count-1] = i2c_readNak();
+        i2c_stop();
+        nunchuck_online();
+    } else {
+        nunchuck_offline();
+    }
+}
+
+typedef enum {
+    SYNC_CONTROLLER,
+    TATACON_PASSTHROUGH
+} State_t;
+
+State_t state = SYNC_CONTROLLER;
+#define ECHOES 2
+int echoes = 0;
+int report_count = 0;
+USB_JoystickReport_Input_t last_report;
+
+void GetNextReport(USB_JoystickReport_Input_t* const reportData) {
+    uint8_t data, i;
+    // Prepare an empty report
+    memset(reportData, 0, sizeof(USB_JoystickReport_Input_t));
+    reportData->LX = STICK_CENTER;
+    reportData->LY = STICK_CENTER;
+    reportData->RX = STICK_CENTER;
+    reportData->RY = STICK_CENTER;
+    reportData->HAT = HAT_CENTER;
+
+    // Repeat ECHOES times the last report
+    if (echoes > 0) {
+        memcpy(reportData, &last_report, sizeof(USB_JoystickReport_Input_t));
+        echoes--;
+        return;
+    }
+
+    // States and moves management
+    switch (state) {
+        case SYNC_CONTROLLER:
+            if (report_count > 100) {
+                report_count = 0;
+                state = TATACON_PASSTHROUGH;
+            }
+            else if (report_count == 25 || report_count == 50) {
+                reportData->Button |= SWITCH_L | SWITCH_R;
+            }
+            else if (report_count == 75 || report_count == 100) {
+                reportData->Button |= SWITCH_A;
+            }
+            report_count++;
+            break;
+        case TATACON_PASSTHROUGH:
```

Comment:
Something really fishy around here.

Pads were triggering a lot of buttons, including Home, which shouldn't happen at all.
Hard for me to catch the source and find a workaround without a proper debugger... :( 

Also, I have almost always double triggering, which should be because of my Tatacon, but debounce (set to 4000 to check if anything happen) seems to do nothing....

Small and dirty workaround I did for the button (didn't found for the debounce yet):
```suggestion
case TATACON_PASSTHROUGH:
            // Get tatacon button data
            data = nunchuck_readByte(BUTTONS_DATA);
            CLEAR(LED_PORT, KAT_LED_PIN);
            CLEAR(LED_PORT, DON_LED_PIN);

            // Tatacon has 4 inputs to check
            for(i = 0; i < TATACON_SWITCHES; i++) {
                // The I2C data starts at the 6th bit and goes down
                uint8_t newState = !(data & _BV(TATACON_BUTTONS_START - i));
                if(!switches[i].debounce && switches[i].state != newState) {
                    switch(i) {
                        case 0 :
                        reportData->Button |= SWITCH_LCLICK;
                        SET(LED_PORT, DON_LED_PIN);
                        break;
                        case 1 :
                        reportData->Button |= SWITCH_ZL;
                        SET(LED_PORT, KAT_LED_PIN);
                        break;
                        case 2 :
                        reportData->Button |= SWITCH_RCLICK;
                        SET(LED_PORT, DON_LED_PIN);
                        break;
                        case 3 :
                        reportData->Button |= SWITCH_ZR;
                        SET(LED_PORT, KAT_LED_PIN);
                        break;
                    }
                    switches[i].state = newState;
                    switches[i].debounce = tataConfig.debounce;
                    switchesChanged = 1;
                }
            }
            break;
```